### PR TITLE
New cryptographic scheme with selective disclosure (Redactable Signature Scheme)

### PR DESCRIPTION
**Pull request by eeelllccc**
## Summary
The addition of a new cryptographic scheme (Redactable Signature Scheme, RSS) providing selective disclosure functionality. The RSS algorithms were implemented from a paper by Olivier Sanders, [Improving Revocation for Group Signature with Redactable Signature](https://eprint.iacr.org/2020/856.pdf), and the [RSS implementation](https://github.com/alan-turing-institute/RSS/tree/new_rss) was motivated by papers by Kumara et al. [[1]](https://arxiv.org/pdf/2310.15934v1.pdf) and [[2]](https://arxiv.org/abs/2310.17297).

## Description
A new `JWK` algorithm (`RSS2023`) and implementation of `ProofSuite` has been used to bring selective disclosure to `LinkedDataDocument`. Selective disclosure has been fully implemented for verifiable credentials in `ssi-vc`, and tested in an integration with [Trustchain](https://github.com/alan-turing-institute/trustchain).

## Crates affected
**`ssi-json-ld`**
- The small change in this crate adjusts the linked data `expansion_policy: json_ld::expansion::Policy::Strict` to `Policy::Relaxed`.

  :::info
  It is intended that this change is only **temporary**, and may be reverted if/when a context is available for the new RSS proof type.
  :::
    - The change makes two unit tests fail: one in the `did-pkh` crate, and the other in the `ssi-vc` crate.
    
**`ssi-jwk`**
- New `rss.rs` module added which impliments a new JWK scheme, following the impl conventions found in the existing work.
- Additions to `lib.rs` to extend the `Algorithm` enum and `get_algorithm` function to include the new `RSS2023` varient.
- A public function for JWK key pair creation, `pub fn generate_rss2023(..)`, is defined and unit tested in `lib.rss`.
- A transparent error variant has been added to `ssi-jwk::Error` which wraps `RSSKeyError` defined in the `rss.rs` mod.
- The byte encoding and decoding of RSS keys are delegated to the `ps_sig` crate which contains the full RSS implementation.

**`ssi-ldp`**
- A new `ProofSuiteType` varient has been added: `RSSSignature2023`.
- The `ProofSuiteType` impls have been extended to cover the new enum varient.
- Most notably, `impl ProofSuite for ProofSuiteType` has been extended with implementations for `ProofSuite::sign(..)` and `ProofSuite::verify(..)`, leaving `ProofSuite::prepare()` `unimplemented!()` for the new `RSSSignature2023` varient.
- The sign and verify implementations are factored out into a new `rss.rs` module in the `suites` directory of the crate.
- The **sign** algorithm:
    - Convert the `LinkedDataDocument` to an RDF dataset, and then into N-Quads.
    - Sort the N-Quads (the sorting method must sort `char`-for-`char` from left the right, such that each Quad is sorted based on the "key" of the information before the "value", and the same sort must be used in the sign and verify algorithms).
    - Map the sorted N-Quads into the `ps_sig::FieldElement` type, a slice of which is the necessary type for the RSS signature signing material.
    - Generate a `ps_sig::RSignature`.
    - Return a new `Proof` with a `ps_sig::RSignature` as the `Proof.value`.
- The **verify** algorithm:
    - Resolves the verification method on the proof.
    - Converts the verification method into a JWK public key.
    - Prepares the verification material, which includes:
        - The public key.
        - The `RSignature` to be verified.
        - A slice of `ps_sig::FieldElement`, which are derived from sorted N-Quads following the procedure in the sign algorithm.
        - A slice of indicies, indicating the positions in the `&[FieldElement]` at which information has been disclosed. In the traditional setting with full disclosure (rather than the option for selective disclosure offered by the RSS scheme), all of the indicies are included.
            ::: info
            Note: The RSS implementation uses **1-based indexing**, so the indicies for a fully disclosed message of length 3 would be **`&[1,2,3]`**.
            :::
    - Inferring the indicies from the document at verification time is non-trivial, but is crucial for alignment with the existing API (the alternative is to require that the indicies are passed in by the document holder and propagated through the API to the verification algorithm).
    - The approach taken to infer which parts of the document have been selectivly disclosed (which indices of the sorted N-Quads representation of the document are included) are as follows:
        - Require that partially disclosed documents retain **all** of the fields of the fully disclosed documents, where the undisclosed fields have their value set to `null` (deserializing to `serde_json::Value::Null`).
        - Create a copy of the document and convert it to an N-Quads representation. This conversion will **drop** `null` fields and produce a **subset** of original N-Quads set that would be produced from a fully disclosed document.
        - Edit a seperate copy of the document to "mark" all `null` values and give them a non-null dummy value.
          ::: warning
          At this stage an `ssi-vc::Credential` specific step is included to remove the `proof` field from the json map. This is not ideal given that the `ssi-ldp` crate should generalise above the `ssi-vc` crate. A suggestion is included in the source code to avoid this.
          :::
        - Producing **sorted** N-Quads from this document without `serde_json::Value::Null` values will result in a **model** of the sorted N-Quads produced from a fully disclosed document.
        - Due to the nature of the sort described above, the model will have N-Quads **in the same order** as the sorted N-Quads that would be produced from a fully disclosed document.
            :::warning
            This only holds true if there are **no arrays** present in the `LinkedDataDocument`. This is one of the major constraints of this implementation.
        - Iterating through the model N-Quads and checking for membership in the subset N-Quads (containing only the disclosed information) reveals the indicies of the disclosed information.
        
    - The signature verification is delegated to the `ps_sig` crate once the verification material has been prepared.
- Signing and verification errors are captured by the `RSSError` enum.
- A new transparent `ssi-ldp::Error::RSS` varient wraps `RSSError`.

**`ssi-vc`**
- `rss_redact(..)` method add to `impl Credential`.
    - Selectively disclose a subset of the credential subject fields whilst maintaining verifiability (applicable to Verifiable Credentials with an RSSSignature2023 type proof). A copy of the credential (with or without proofs included) is passed in with a partially masked `CredentialSubject` json map. Fields are masked by setting them to `serde_json::value::Value::Null`.
    - The `ps_sig` crate is used to generate a new "derived" signature which verifies successfully against a subset of the information in the credential.
    - The algorithm is as follows:
        1. Carry out 2 checks for consistency between the `Credential` document (`self`) and the `masked_copy` input argument.
        2. Infer the indicies corresponding to the disclosed information in the masked copy of the document.
        3. Pass the indicies and a flattened version of the **original** document (`self`) as signing input to the RSS signature derivation.
        4. Add the new derived proof to the masked_copy credential passed in to this method.
        5. Modify `self` by replacement with the signed masked_copy credential.
- Unit tests cover issuance, verification, redaction and the handling of several error scenarios where the `masked_copy` of a credential, passed in by a holder, is corrupt.

**`ssi-dids`**
- Addition of a new `did:example:rss` DID to the example resolver to facilitate unit tests in other crates.

**`ssi-jwt`**
- A minor change required for integration with Trustchain.

### CODE REVIEW

File: ssi-json-ld/src/lib.rs [hunk 1]
```diff
@@ -520,7 +520,7 @@ where
         expand_context,
         // VC HTTP API Test Suite expect properties to not be silently dropped.
         // More info: https://github.com/timothee-haudebourg/json-ld/issues/13
-        expansion_policy: json_ld::expansion::Policy::Strict,
+        expansion_policy: json_ld::expansion::Policy::Relaxed,
```
Comment (eeelllccc) [inline ssi-json-ld/src/lib.rs:523]: Make a note about this in the PR description.

File: ssi-jwk/Cargo.toml [hunk 2]
```diff
@@ -69,6 +71,7 @@ multibase = "0.9.1"
 unsigned-varint = "0.7.1"
 num-traits = "0.2"
 num-derive = "0.3"
+ps_sig = { git = "https://github.com/alan-turing-institute/RSS.git", branch = "new_rss", optional = true }
```
Comment (eeelllccc) [inline ssi-jwk/Cargo.toml]: Point to a commit here? Also what to do about lots of warnings generated by the ps_sig crate (warning generated by existing ps_sig code).

File: ssi-jwt/src/lib.rs [hunk 3]
```diff
@@ -16,7 +16,6 @@ pub fn encode_sign<Claims: Serialize>(
     let header = Header {
         algorithm,
         key_id: key.key_id.clone(),
-        type_: Some("JWT".to_string()),
```
Comment (eeelllccc) [inline ssi-jwt/src/lib.rs:19]: Not sure why this was done. Test without deletion.

File: ssi-ldp/src/suites/mod.rs [hunk 4]
```diff
@@ -190,6 +198,8 @@ impl ProofSuiteType {
             #[cfg(feature = "secp256r1")]
             Self::EcdsaSecp256r1Signature2019 => &["https://w3id.org/security#EcdsaSecp256r1Signature2019"],
             Self::CLSignature2019 => todo!(),
+            #[cfg(feature = "rss")]
+            Self::RSSSignature2023 => &[],
```
Comment (eeelllccc) [inline ssi-ldp/src/suites/mod.rs:202]: Replace with an unimplemented!() ?

File: ssi-ldp/src/suites/rss.rs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                set.insert(q);
+                set
+            });
+
+    // Mark null values in a json Value
+    fn mark_null(val: Value, null_marker: &str) -> Value {
+        match val {
+            Value::Null => Value::String(null_marker.to_string()),
+            Value::Object(mut map) => {
+                // TODO: this is specific to credentials, and this crate should generalise above them
```
Comment (eeelllccc) [inline ssi-ldp/src/suites/rss.rs]: Add note in PR description

File: ssi-ldp/src/suites/rss.rs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+                if map.contains_key("proof") {
+                    map.remove("proof").unwrap();
+                }
+                Value::Object(
+                    map.into_iter()
+                        .map(|(k, v)| (k, mark_null(v, NULL_MARKER)))
+                        .collect(),
+                )
+            }
+            // TODO: Arrays are not handled here, throw Err if encountered?
```
Comment (eeelllccc) [inline ssi-ldp/src/suites/rss.rs:160]: Add note in PR description

File: ssi-ldp/src/suites/rss.rs [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        }
+    }
+}
+
+pub struct InferredDataset {
+    pub inferred_idxs: Vec<usize>,
+    pub null_marked_dataset_quads: Vec<String>,
+}
+
+const NULL_MARKER: &str = "__12345__";
```
Comment (eeelllccc) [inline ssi-ldp/src/suites/rss.rs:133]: Is there a test to prove that the algorithm is insensitive to the choice of null marker/ whether or not the null marker appears as a value in the credential?

File: ssi-vc/src/lib.rs [hunk 8]
```diff
@@ -1086,6 +1089,165 @@ impl Credential {
...
+
+        // Helper function to set all values in a json map to Null (not supporting Arrays).
+        fn value_to_null(val: &mut Value) {
+            match val {
+                Value::Null => {}
+                Value::Bool(_) | Value::String(_) | Value::Number(_) => *val = Value::Null,
+                Value::Object(ref mut map) => {
+                    map.iter_mut().for_each(|(_, v)| value_to_null(v));
+                }
+                // TODO: Arrays are not handled here, throw Err if encountered?
```
Comment (edchapman88) [inline ssi-vc/src/lib.rs:1116]:
```suggestion
```

File: ssi-vc/src/lib.rs [hunk 9]
```diff
@@ -1086,6 +1089,165 @@ impl Credential {
         result.checks.push(Check::Status);
         result
     }
+
+    /// Selectively disclose a subset of the credential subject fields whilst maintaining
+    /// verifiability (applicable to Verifiable Credentials with a an RSSSignature2023 type proof). A
```
Comment (eeelllccc) [inline ssi-vc/src/lib.rs]:
```suggestion
    /// verifiability (applicable to Verifiable Credentials with an RSSSignature2023 type proof). A
```

File: ssi-jwt/src/lib.rs (hunk 3 above)
Comment (eeelllccc) [reply ssi-jwt/src/lib.rs:19]: This is the one change made on the "modify-encode-sign" branch before the RSS work started.

**Pull request merged**